core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java

```
@@ -925,34 +933,157 @@ public boolean isHashValid(String hash) {
+        public static final int KEY_LENGTH = 512;
+        public static final String PBKDF2_ALGORITHM = "PBKDF2WithHmacSHA512";
+
+        @SuppressFBWarnings(value = "MS_SHOULD_BE_FINAL", justification = "Accessible via System Groovy Scripts")
+        @Restricted(NoExternalUse.class)
+        private static int MAXIMUM_PBKDF2_LOG_ROUND = SystemProperties.getInteger(HudsonPrivateSecurityRealm.class.getName() + ".maximumPBKDF2LogRound", 1000);
+
+        private static final Pattern PBKDF2_PATTERN = Pattern.compile("^\\$PBKDF2\\$HMACSHA512\\:([0-9]{4})\\:[a-zA-Z0-9=]+\\$[a-zA-Z0-9=]+");
+
+        @Override
+        public String encode(CharSequence rawPassword) {
+            return generatePasswordHashWithPBKDF2(rawPassword);
+        }
+
+        @Override
+        public boolean matches(CharSequence rawPassword, String encodedPassword) {
+                return validatePassword(rawPassword.toString(), encodedPassword);
+        }
+
+        private static String generatePasswordHashWithPBKDF2(CharSequence password)  {
+            byte[] salt = generateSalt();
+            PBEKeySpec spec = new PBEKeySpec(password.toString().toCharArray(), salt, 1000, KEY_LENGTH);
+            byte[] hash = generateSecretKey(spec);
+            return PBKDF2 + "$HMACSHA512:" + 1000 + STRING_SEPARATION + Util.toHexString(salt) + "$" + Util.toHexString(hash);
+        }
+
+        private static byte[] generateSecretKey(PBEKeySpec spec)  {
+            SecretKeyFactory secretKeyFactory = null;
+            byte[] hash = new byte[0];
+            try {
+                secretKeyFactory = SecretKeyFactory.getInstance(PBKDF2_ALGORITHM);
+                hash = secretKeyFactory.generateSecret(spec).getEncoded();
+            } catch (NoSuchAlgorithmException e) {
+                LOGGER.log(WARNING, "No such algorithm found for encode",  e);
+            } catch (InvalidKeySpecException e) {
+                LOGGER.log(WARNING, "Invalid key spec exception",  e);
+            }
```

so this returns an empty (zero length) byte array if we can not get the algotithm or the keysepc is invalid?

There are 2 things here

1. this is a silent failure as far as a user attempting to login - things could just fail with no indication things went wrong.
2. how does this impact the security - are we not going to take *any* supplied password and encode it to the same empty zero lenght byte array which means you have no security?  this would seem like an exceptionally bad security issue if true.

---

modified to throw exception